torch/_torch_docs.py

```diff
@@ -3152,7 +3152,16 @@ def merge_dicts(*dicts):
            r"""
 flatten(input, start_dim=0, end_dim=-1) -> Tensor
 
-Flattens a contiguous range of dims in a tensor.
+Returns tensor with the same data and number of elements as :attr:`input`,
+but with dimensions in [:attr:`start_dim`, :attr:`end_dim`] collapsed into a
+single dimension. When possible, the returned tensor will be a view of :attr:`input`.
+Otherwise, it will be a copy.
+
+See :meth:`torch.Tensor.view` on when it is possible to return a view.
```

Comment:
This is a great reference to view to avoid doc redundancy, but I wonder if we can be clearer here. Is there some way to express what view operation this is equivalent to when the result shape can view the input tensor, and what operation it's equivalent to when the result shape can't view the input tensor?

Reply:
For example, after the first sentence something like this might work:

If `start_dim != end_dim`, then shape of :attr:\`input\` can be described as `(l*, f+, r*)`, where `l*` is the sizes of the zero or more "left" dimensions, `f+` is the sizes of the one or more dimensions between :attr:\`start_dim\` and :attr:\`end_dim\` that will be "flattened", and `r*` is the sizes of the zero or more "right" dimensions in :attr:\`input\`. The flattened shape is `(l*, \prod f+, r*)`, and the operation is equivalent to calling `torch.reshape(t, (l*, \prod f+, r*))`. 

Like :func:\`torch.reshape\`, when `start_dim != end_dim` flatten returns a view when possible. See :meth:\`torch.Tensor.view` for details.

If `start_dim == end_dim` then this returns :attr:\`input\` when :attr:\`input\` has one or more dimensions, and a one-dimensional view of :attr:\`input` when :attr:\`input\` has zero dimensions. 


I think these three paragraphs can be improved, but I believe they capture the details of flatten's behavior. Also note that flatten may return self, a view, or a copy. See here for returning self:

https://github.com/pytorch/pytorch/blob/6f928a4a53f022c486df6416e73c69e2acfdd2de/aten/src/ATen/native/TensorShape.cpp#L1639

A test for this case should be added to view ops, too.

Reply:
- Rearranged first paragraph so that the goal of describing input shape->output shape is more clear from the get-go. 
 - Kept zero-dimensional case factored out into a note.